### PR TITLE
kfctl uses kfconfig

### PR DESCRIPTION
**Pull request by lluunn**
/cc @gabrielwen @jlewi

Related to #3703

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/4250)

### CODE REVIEW

File: bootstrap/pkg/kfapp/aws/aws.go [hunk 1]
```diff
@@ -106,6 +107,10 @@ func (aws *Aws) GetK8sConfig() (*rest.Config, *clientcmdapi.Config) {
 	return nil, nil
 }
 
+func (aws *Aws) GetKfConfig() *kfconfig.KfConfig {
```
Comment (jlewi) [inline bootstrap/pkg/kfapp/aws/aws.go]: Why did the function signature of this method change? The original method was returning *rest.Config and clientcmdapi.Cfg?

File: bootstrap/pkg/kfapp/coordinator/coordinator.go [hunk 2]
```diff
@@ -108,11 +109,21 @@ func getConfigFromCache(pathDir string, kfDef *kfdefsv3.KfDef) ([]byte, error) {
 // GetPlatform will return an implementation of kftypesv3.GetPlatform that matches the platform string
 // It looks for statically compiled-in implementations, otherwise throws unrecognized error
 func getPlatform(kfdef *kfdefsv3.KfDef) (kftypesv3.Platform, error) {
+	kfdefBytes, err := yaml.Marshal(kfdef)
```
Comment (jlewi) [inline bootstrap/pkg/kfapp/coordinator/coordinator.go]: Why is this function taking a kfDef  and not a KfConfig?

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 3]
```diff
@@ -117,6 +118,10 @@ type Setter interface {
 	SetRunGetCredentials(v bool)
 }
 
+func (gcp *Gcp) GetKfConfig() *kfconfig.KfConfig {
```
Comment (jlewi) [inline bootstrap/pkg/kfapp/gcp/gcp.go]: Why the new method?

File: bootstrap/pkg/kfapp/gcp/gcp.go [hunk 4]
```diff
@@ -216,8 +222,8 @@ func (gcp *Gcp) GetK8sConfig() (*rest.Config, *clientcmdapi.Config) {
 	// TODO(jlewi): Should we fix this so we can build a cluster config which takes
 	// a TokenSource which can then be pointed at either the DefaultTokenSource
 	// or the refreshable token source?
-	restConfig, err := utils.BuildClusterConfig(ctx, accessToken, gcp.kfDef.Spec.Project,
-		gcp.kfDef.Spec.Zone, gcp.kfDef.Name)
+	restConfig, err := utils.BuildClusterConfig(ctx, accessToken, gcp.kfDef.Project,
```
Comment (jlewi) [inline bootstrap/pkg/kfapp/gcp/gcp.go]: Why isn't it gcp.KfDef.Spec.Project? The whole point of KfConfig is that it should be a drop in replacement.

File: bootstrap/pkg/apis/apps/configconverters/v1alpha1.go [hunk 5]
```diff
@@ -61,13 +61,14 @@ func (v V1alpha1) ToKfConfig(appdir string, kfdefBytes []byte) (*kfconfig.KfConf
 	}
 
 	config := &kfconfig.KfConfig{
```
Comment (jlewi) [inline bootstrap/pkg/apis/apps/configconverters/v1alpha1.go]: Why doesn't KFConfig have a Spec and a Status?

**Comment by jlewi** [comment]
It looks like KFConfig needs some changes so it can be a drop in replacement for KFDef. Specifically KFConfig should have a Spec and Status.

This way
KFDef.Spec.Project  can just be replaced with KFConfig.Spec.Project

File: bootstrap/pkg/kfapp/coordinator/coordinator.go (hunk 2 above)
Comment (lluunn) [reply bootstrap/pkg/kfapp/coordinator/coordinator.go]: We should eventually make it take kfconfig after all platforms change to take kfconfig.
But for now, other platforms are taking a kfdef

File: bootstrap/pkg/kfapp/gcp/gcp.go (hunk 3 above)
Comment (lluunn) [reply bootstrap/pkg/kfapp/gcp/gcp.go]: So outside can access kfconfig. any conern?

File: bootstrap/pkg/kfapp/gcp/gcp.go (hunk 4 above)
Comment (lluunn) [reply bootstrap/pkg/kfapp/gcp/gcp.go]: done

File: bootstrap/pkg/apis/apps/configconverters/v1alpha1.go (hunk 5 above)
Comment (lluunn) [reply bootstrap/pkg/apis/apps/configconverters/v1alpha1.go]: done

File: bootstrap/pkg/kfapp/aws/aws.go (hunk 1 above)
Comment (lluunn) [reply bootstrap/pkg/kfapp/aws/aws.go]: I don't follow, this is to provide a method to access kfconfig.
What do you mean by change?

**Comment by jlewi** [comment]
On my end I think this is ready to go once tests pass and conflicts are resolved. I had a couple minor questions.

@gabrielwen could you please review?

**Comment by lluunn** [comment]
current error: 
```
Error from server (Forbidden): error when creating "/tmp/kout159894712": persistentvolumes "minio-pv" is forbidden: err
or querying GCE PD volume test1-storage-artifact-store: disk is not found
```

**Comment by jlewi** [comment]
@lluunn I suspect that is an issue with using deployment manager to create a PD that will back the volume. I think this is some of the relevant code
https://github.com/kubeflow/kubeflow/blob/321c1d3313b6b898a64616d3b54e7db2179d89d6/bootstrap/pkg/kfapp/gcp/gcp.go#L1328

My guess is that we didn't actually create the PD but we ended up configuring pipelines to use the PD. 

I think this is the overlay that configures minio PVC to be backed a PV.
https://github.com/kubeflow/manifests/tree/master/pipeline/minio/overlays/minioPd

So my guess is there is an inconsistency somewhere; i.e. we added the overlay but didn't actually use the DM config to create the PD.

**Comment by gabrielwen** [comment]
@jlewi yea I'm working with lunkai to finish this PR.

**Comment by gabrielwen** [comment]
@lluunn could you print out this value to see if we convert the KfConfig correctly?
https://github.com/kubeflow/kubeflow/blob/321c1d3313b6b898a64616d3b54e7db2179d89d6/bootstrap/pkg/kfapp/gcp/gcp_plugin.go#L104:25

**Comment by lluunn** [comment]
I think it's the name that's incorrect.
I see there is a deployment for storage, but the name should be kfctl-12e7, and the manifest is looking for name test1

**Comment by lluunn** [comment]
The problem seems to be that in gcp.Generate we fill in the correct params, but in kustomize.Generate it overwrites it.

**Comment by lluunn** [comment]
failure:
workflow=kubeflow-presubmit-kfctl-go-basic-auth-4250-9984c15-3728-82fc
failed at onExit (kfctl delete)

klon kubeflow-test-infra kubeflow-presubmit-kfctl-go-basic-auth-4250-9984c15-3728-82fc-3589979666 main

```
Endpoint is not deleted: kfctl-926d.endpoints.kubeflow-ci-deployment.cloud.goog
FAILED

=================================== FAILURES ===================================
______________________________ test_kfctl_delete _______________________________

kfctl_path = '/mnt/test-data-volume/kubeflow-presubmit-kfctl-go-basic-auth-4250-9984c15-3728-82fc/src/kubeflow/kubeflow/bootstrap/bin/kfctl'
app_path = '/mnt/test-data-volume/kubeflow-presubmit-kfctl-go-basic-auth-4250-9984c15-3728-82fc/kfctl-926d'
project = 'kubeflow-ci-deployment'

    def test_kfctl_delete(kfctl_path, app_path, project):
      if not kfctl_path:
        raise ValueError("kfctl_path is required")

      if not app_path:
        raise ValueError("app_path is required")

      logging.info("Using kfctl path %s", kfctl_path)
      logging.info("Using app path %s", app_path)

      util.run([kfctl_path, "delete", "--delete_storage", "-V"],
               cwd=app_path)

      # Use services.list instead of services.get because error returned is not
      # 404, it's 403 which is confusing.
      name = os.path.basename(app_path)
      endpoint_name = "{deployment}.endpoints.{project}.cloud.goog".format(
          deployment=name,
          project=project)
      logging.info("Verify endpoint service is deleted: " + endpoint_name)
      if endpoint_name in get_endpoints_list(project):
        msg = "Endpoint is not deleted: " + endpoint_name
        logging.error(msg)
>       raise AssertionError(msg)
E       AssertionError: Endpoint is not deleted: kfctl-926d.endpoints.kubeflow-ci-deployment.cloud.goog
```

and 

```
util.py                     71 INFO     time="2019-10-10T03:22:26Z" level=info msg="Removing serviceAccount:kfctl-926d-user@kubeflow-ci-deployment.iam.gserviceaccount.com from roles/viewer" filename="gcp/gcp.go:1021"
util.py                     71 INFO     time="2019-10-10T03:22:29Z" level=info msg="Endpoint  deletion is failed but it is not found in managed services, treating it as successful." filename="gcp/gcp.go:931"
kfctl_delete_test.py        59 INFO     Verify endpoint service is deleted: kfctl-926d.endpoints.kubeflow-ci-deployment.cloud.goog
__init__.py                 44 WARNING  file_cache is unavailable when using oauth2client >= 4.0.0
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/googleapiclient/discovery_cache/__init__.py", line 36, in autodetect
    from google.appengine.api import memcache
ImportError: No module named 'google.appengine'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/googleapiclient/discovery_cache/file_cache.py", line 33, in <module>
    from oauth2client.contrib.locked_file import LockedFile
ImportError: No module named 'oauth2client.contrib.locked_file'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/googleapiclient/discovery_cache/file_cache.py", line 37, in <module>
    from oauth2client.locked_file import LockedFile
ImportError: No module named 'oauth2client.locked_file'
```

**Comment by lluunn** [comment]
test all passed now!
but need to fix conflict..

**Comment by gabrielwen** [comment]
will take a pass first thing on Monday!  Thanks for finishing this!!!

**Comment by jlewi** [comment]
Woo Hoo!

**Comment by jlewi** [comment]
@lluunn @gabrielwen Can the changes under
apps/configconverters  and
kfconfig

be split off into their own PR? 

IIUC no code depends on KFConfig yet so those changes should be able to be submitted without needing to update a whole bunch of other code. Those changes should also be fairly straightforward to review.

That would make it a lot easier to focus on the changes to coordinator.go/aws.go/etc...

**Comment by jlewi** [comment]
@lluunn @gabrielwen Thank you both very much for this!

My high level comment is can we split off some of the more straightforward changes into small PRs that we merge first? This will make the code review go much faster. Looking at the code I see potentially 3 small PRs that could be merged first that would drastically reduce the size and complexity of this PR

1. All the updates to apps/configconverters and kfconfig
1. Removal of  dead code from kustomize.go
    * i.e. initComponents
1. Deletion of minikube.go

**Comment by yanniszark** [comment]
> @yanniszark Embedding KFDef/KFConfig inside Existing seems like its going to be an issue. I thought we had changed that to always be a pointer.

@jlewi agreed. Thanks for pointing it out!
Filed https://github.com/kubeflow/kubeflow/issues/4293

> KFDef.Spec.Project can just be replaced with KFConfig.Spec.Project

It seems kind of weird to me to have a plugin-specific field (project) inside the spec.
Shouldn't it be inside the gcp plugin's spec? Am I missing something here?

**Comment by gabrielwen** [comment]
> > @yanniszark Embedding KFDef/KFConfig inside Existing seems like its going to be an issue. I thought we had changed that to always be a pointer.
> 
> @jlewi agreed. Thanks for pointing it out!
> Filed #4293
> 
> > KFDef.Spec.Project can just be replaced with KFConfig.Spec.Project
> 
> It seems kind of weird to me to have a plugin-specific field (project) inside the spec.
> Shouldn't it be inside the gcp plugin's spec? Am I missing something here?

this is for backward compatibility - you are right it's plugin specific but currently KfDef.Spec.Project is shared with coordinator/kustomize

**Comment by jlewi** [comment]
@lluunn it looks like you may have done a merge of master into your branch rather than a rebase. That will likely cause problems. 

You'll probably need to fix that at some point. I might suggest something like the following

Create a new local branch based on your local branch gcpkfconfigs (so that if you mess up you can just start over easily)

```
git checkout -b gcpkfconfigs_fix
```

Now do an interactive rebase on master

```
git rebase -i upstream/master
```

I think you'll want to drop all the commits corresponding to merges.

You'll undoubtedly get a bunch of conflicts but I wouldn't even bother trying to resolve them just continue (i.e. commit the files even if the conflict markers) inside them.

Once the rebase is done your files will have a bunch of conflict markers in them but that's ok you can resolve this by overwriting the files on gcpkfconfigs_fix with the files on the tip of your gcpkfconfig1 branch. e.g.

```
git checkout gcpkfconfig1 bootstrap/pkg/kfapp/coordinator.go
```

You'll probably want to do a diff against upstream/master and restore any changes on master that got dropped. (I usually use [meld](https://stackoverflow.com/questions/34119866/setting-up-and-using-meld-as-your-git-difftool-and-mergetool) for this)

```
git difftool -d upstream/master 
```

If everything looks good on your local branch gcpkfconfig1_new then you can force push it to the remote branch gcpkfconfig1 to update the PR.

**Comment by lluunn** [comment]
Thanks @jlewi .
I followed the steps, but just pushed to a new branch to be safe: #4307.

**Comment by jlewi** [comment]
@lluunn I'd like to continue to use this PR so we preserve the review history. I can help you when I get into the office to update this safely but if you have 2 branches locally one that was your origiinal branch and one that is the new branch with all the rebases (i.e. what you pushed to #4307)  then you can safely force push the remote branch lluun:gcpkfconfig1. And you can always recover from your local branches.

Even then when you rebase those changes are preserved and you can do `git reflog` to see that history and go back. 

If your truly paranoid just do `cp -r` your local clone repo directory to create a backup.

File: bootstrap/pkg/kfapp/coordinator/coordinator.go [hunk 6]
```diff
@@ -385,29 +337,30 @@ func LoadKfAppCfgFile(cfgfile string) (kftypesv3.KfApp, error) {
 	// Set some defaults
 	// TODO(jlewi): This code doesn't belong here. It should probably be called from inside KfApp; e.g. from
 	// KfApp.generate. We should do all initialization of defaults as part of the reconcile loop in one function.
-	if c.KfDef.Spec.PackageManager == "" {
-		c.KfDef.Spec.PackageManager = kftypesv3.KUSTOMIZE
-	}
+	// if c.KfDef.Spec.PackageManager == "" {
```
Comment (jlewi) [inline bootstrap/pkg/kfapp/coordinator/coordinator.go]: Why did you comment this out?

File: bootstrap/pkg/kfapp/minikube/minikube.go [hunk 7]
```diff
@@ -60,56 +61,58 @@ func (minikube *Minikube) Delete(resources kftypes.ResourceEnum) error {
 }
 
 func (minikube *Minikube) generate() error {
```
Comment (jlewi) [inline bootstrap/pkg/kfapp/minikube/minikube.go]: @lluunn @gabrielwen can one of you just file a separate PR to delete the minikube code?

File: bootstrap/pkg/kfupgrade/kfupgrade.go [hunk 8]
```diff
@@ -216,17 +216,21 @@ func MergeKfDef(oldKfDef *kfdefsv3.KfDef, newKfDef *kfdefsv3.KfDef) {
 }
 
 func (upgrader *KfUpgrader) Generate() error {
-	kfApp := kustomize.GetKfApp(upgrader.NewKfDef)
-	return kfApp.Generate(kftypesv3.K8S)
+	// TODO: fix this
```
Comment (jlewi) [inline bootstrap/pkg/kfupgrade/kfupgrade.go]: Why is this commented out?

Comment (jlewi) [reply bootstrap/pkg/kfupgrade/kfupgrade.go]: @richardsliu Is this a bug in kfupgrader? Why are we calling kustomize.GetKFApp?

File: bootstrap/pkg/kfapp/coordinator/coordinator.go (hunk 6 above)
Comment (lluunn) [reply bootstrap/pkg/kfapp/coordinator/coordinator.go]: we don't need packagemanager field right?
It's always kustomize.

File: bootstrap/pkg/kfapp/minikube/minikube.go (hunk 7 above)
Comment (lluunn) [reply bootstrap/pkg/kfapp/minikube/minikube.go]: will do

File: bootstrap/pkg/kfupgrade/kfupgrade.go (hunk 8 above)
Comment (lluunn) [reply bootstrap/pkg/kfupgrade/kfupgrade.go]: I need to either migrate kfupgrade to use kfconfig as well in this PR, or in a followup PR.
WDYT?

Comment (jlewi) [reply bootstrap/pkg/kfupgrade/kfupgrade.go]: Chatted with @richardsliu He's going to update KfUpgrader. I think what we discussed is

* kfupgrader will be updated to store the filepath of the KFDef
* We will call coordinator.
https://github.com/kubeflow/kubeflow/blob/9f5ef66ce3d12ff2b603d07330a16661caa0b286/bootstrap/pkg/kfapp/coordinator/coordinator.go#L220

**Comment by gabrielwen** [comment]
test was failed because we've reached the cap for iam policies.

**Comment by gabrielwen** [comment]
```
util.py                     71 INFO     Error: internal error: coordinator does not implement KfDefGetter                                                                          
util.py                     71 INFO     Usage:                                                                                                                                     
util.py                     71 INFO     kfctl delete [flags]                                                                                                                       
util.py                     71 INFO                                                                                                                                                
util.py                     71 INFO     Flags:                                                                                                                                     
util.py                     71 INFO     --delete_storage   Set if you want to delete app's storage cluster used for mlpipeline.                                                    
util.py                     71 INFO     -h, --help             help for delete                                                                                                     
util.py                     71 INFO     -V, --verbose          verbose output default is false                                                                                     
util.py                     71 INFO                                                                                                                                                
util.py                     71 INFO     internal error: coordinator does not implement KfDefGetter
```

File: bootstrap/pkg/kfupgrade/kfupgrade.go (hunk 8 above)
Comment (lluunn) [reply bootstrap/pkg/kfupgrade/kfupgrade.go]: so ok if I leave it as commented out here?

Comment (jlewi) [reply bootstrap/pkg/kfupgrade/kfupgrade.go]: I think so.

**Comment by lluunn** [comment]
resource error: does not have available private IP space in 10.0.0.0/8 to reserve a /14 block

**Comment by lluunn** [comment]
seems delete taking longer than expected:
```
time="2019-10-16T00:04:48Z" level=info msg="Deleting kfctl-d65a status: RUNNING (op = operation-1571183378810-594fba32b
c6f5-49198a7d-1e1aa925)" filename="gcp/gcp.go:386"
util.py                     71 INFO     Error: couldn't delete KfApp:  (kubeflow.error): Code 500 with message: coordinator Delete failed for gcp:  (kubeflow.e
rror): Code 400 with message: Gcp.Delete is failed for kubeflow-ci-deployment/kfctl-d65a: Deleting kfctl-d65a did not succeed; status: RUNNING (op = operation-
1571183378810-594fba32bc6f5-49198a7d-1e1aa925)
```

**Comment by lluunn** [comment]
verified the deployment kfctl-d65a is actually deleted.

File: bootstrap/pkg/kfapp/coordinator/coordinator.go (hunk 6 above)
Comment (swiftdiaries) [reply bootstrap/pkg/kfapp/coordinator/coordinator.go]: +1

**Comment by lluunn** [comment]
test all passed, but need to sync master

**Pull request merged**